contracts/YieldEulerLever.sol

```diff
@@ -0,0 +1,390 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.14;
+
+import "@yield-protocol/utils-v2/contracts/token/TransferHelper.sol";
+import "./interfaces/IEulerMarkets.sol";
+import "./interfaces/IEulerEToken.sol";
+import "./YieldLeverBase.sol";
+
+// Get flash loan of USDC/DAI/WETH
+// Deposit to get eulerToken
+// Deposit & borrow against it
+// Sell the fyToken to get USDC/DAI
+// Close the flash loan
+contract YieldEulerLever is YieldLeverBase {
+    using TransferHelper for IERC20;
+
+    // address constant EULER_MAINNET;
+    address constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+    address constant DAI = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+    address constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address constant EULER = 0x27182842E098f60e3D576794A5bFFb0777E025d3;
+
+    // Use the markets module:
+    IEulerMarkets public constant markets =
+        IEulerMarkets(0x3520d5a913427E6F0D6A83E07ccD4A4da316e4d3);
+
+    mapping(bytes6 => FlashJoin) public flashJoins;
+    mapping(bytes6 => address) public ilkToAsset;
+
+    // EulerMainMarket: 0x3520d5a913427E6F0D6A83E07ccD4A4da316e4d3
+    constructor(Giver giver_) YieldLeverBase(giver_) {
+        // Approve the main euler contract to pull your tokens:
+        IERC20(USDC).approve(EULER, type(uint256).max);
+        IERC20(DAI).approve(EULER, type(uint256).max);
+        IERC20(WETH).approve(EULER, type(uint256).max);
+
+        flashJoins[0x323000000000] = FlashJoin(
+            0x4fE92119CDf873Cf8826F4E6EcfD4E578E3D44Dc
+        ); // daiJoin
+        flashJoins[0x323100000000] = FlashJoin(
+            0x0d9A1A773be5a83eEbda23bf98efB8585C3ae4f4
+        ); // usdcJoin
+        flashJoins[0x323200000000] = FlashJoin(
+            0x3bDb887Dc46ec0E964Df89fFE2980db0121f0fD0
+        ); // wethJoin
+
+        ilkToAsset[0x323000000000] = DAI;
+        ilkToAsset[0x323100000000] = USDC;
+        ilkToAsset[0x323200000000] = WETH;
+
+        // Approve join for
+        IERC20(USDC).approve(
+            0x0d9A1A773be5a83eEbda23bf98efB8585C3ae4f4, // USDC Join
+            type(uint256).max
+        );
+        IERC20(DAI).approve(
+            0x4fE92119CDf873Cf8826F4E6EcfD4E578E3D44Dc, // DAI Join
+            type(uint256).max
+        );
+        IERC20(WETH).approve(
+            0x3bDb887Dc46ec0E964Df89fFE2980db0121f0fD0, // WETH Join
+            type(uint256).max
+        );
+    }
+
+    /// @notice Approve maximally for an fyToken.
+    /// @param seriesId The id of the pool to approve to.
+    function approveFyToken(bytes6 seriesId) external {
+        IPool(ladle.pools(seriesId)).fyToken().approve(
+            address(ladle),
+            type(uint256).max
+        );
+    }
+
+    /// @notice Invest by creating a levered vault.
```

Comment:
It should be clearer in the natspec that the contract will take eTokens from the caller. An ascii diagram of token flows would be nice.